`surv.xgboost` model type split + add `distr` predictions to `surv.xgboost.cox`

bblodfon opened the pull request:
First draft addressing https://github.com/mlr-org/mlr3extralearners/issues/316. Comments:
- Doc templates are not able to distinguish between the two types? (need help with this!)
- Conflict with `objective` parameter, e.g. by running
```
cox = lrn("surv.xgboost.cox", objective = "survival:aft")
```
you get `<LearnerSurvXgboost:surv.xgboost.cox>: Gradient Boosting (Cox)` with `objective:aft`! How to resolve this? Easy way would be to not expose the `objective` parameter to the user and use only the `type`, makes sense?
- Depends on https://github.com/mlr-org/mlr3proba/pull/338 for the `distr` for the `cox` type (TODO)

sebffischer commented on R/learner_xgboost_surv_xgboost.R:
```diff
@@ -12,7 +12,9 @@
 #' - `nthread` is initialized to 1 to avoid conflicts with parallelization via \CRANpkg{future}.
 #' - `verbose` is initialized to 0.
-#' - `objective` is initialized to `survival:cox` for survival analysis.
+#' @section Custom mlr3 parameters:
```
why is this a custom parameter and not an initial value?

bblodfon replied:
the `type` argument is an initialization parameter but mlr3-specific, though it affects one of the parameters (`objective`). I think that it is better to delete the `objective` in this case and specify only the `type` to make the split consistent (see issue in the comment above). What do you think?

sebffischer commented on R/learner_xgboost_surv_xgboost.R:
```diff
@@ -115,17 +119,18 @@ LearnerSurvXgboost = R6Class("LearnerSurvXgboost",
 
       super$initialize(
-        id = "surv.xgboost",
+        id = paste0("surv.xgboost.", type),
```
also update NEWs please for all the fixes / changes :)

be-marc commented:
So when I started, Michel always told me to stick to the [KISS](https://en.wikipedia.org/wiki/KISS_principle) principle. I think that also applies here. The type argument only adds unnecessary complexity. You are also moving away from OOP. If there are conceptually two classes, then make them two classes. Turning one class into two with if statements is rather bad style. If you don't want to repeat code, use helper functions, we do that quite often in our packages.

sebffischer commented:
sorry @bblodfon i did not think of these problems when i suggested the solution. 
I don’t think it is that bad in principle because we save a lot of lines of code with very little logic, but because of the templates i guess it makes sense to have two different learner classes with helper functions as marc suggested. Also i would try to maintain that the ID surv.xgboost still constructs the learner as it used to be, for backwards compatibility.

bblodfon commented:
- original `surv.xgboost` was a bit refactored, but kept intact otherwise
- added 2 new classes/learners: `surv.xgboost.cox` and `surv.xgboost.aft`
- added `helpers_xgboost.R` and a doc template
- add breslow in `surv.xgboost.cox` **TODO**

bblodfon commented:
@sebffischer ready for review!
- My big question is: should we keep the original `surv.xgboost`? (I would vote to completely remove it, at some point)
- Maybe a good idea would be to say something in the doc, eg "better to use `surv.xgboost.cox` and `surv.xgboost.aft` separately, as prediction types now differ from the vanilla `surv.xgboost`

sebffischer commented on R/learner_xgboost_surv_xgboost.R:
```diff
@@ -6,24 +6,22 @@
 #' eXtreme Gradient Boosting regression.
 #' Calls [xgboost::xgb.train()] from package \CRANpkg{xgboost}.
 #'
+#' **Note:** We strongly advise to use the separate [Cox][LearnerSurvXgboostCox]
```
maybe you want to add a deprecated warning to the old xgboost learner?

sebffischer replied:
then at some point later we can remove it as you suggested.

bblodfon replied:
Added one yes

sebffischer commented on R/learner_xgboost_surv_xgboost.R:
```diff
@@ -71,7 +69,6 @@ LearnerSurvXgboost = R6Class("LearnerSurvXgboost",
         normalize_type              = p_fct(c("tree", "forest"), default = "tree", tags = "train"),
         nrounds                     = p_int(1L, tags = "train"),
         nthread                     = p_int(1L, default = 1L, tags = c("train", "threads")),
-        ntreelimit                  = p_int(1L, tags = "predict"),
```
why was this removed?

bblodfon replied:
So in the `predict.XGBoost` method I saw that `iterationrange` is used instead (which we include either way): 
![ntreelimit](https://github.com/mlr-org/mlr3extralearners/assets/10909555/3db1cdcd-4a20-4473-a6d7-837e3964f70a)

sebffischer commented on R/learner_xgboost_surv_xgboost_aft.R:
```diff
@@ -0,0 +1,195 @@
+#' - `nrounds` is initialized to 1.
+#' - `nthread` is initialized to 1 to avoid conflicts with parallelization via \CRANpkg{future}.
+#' - `verbose` is initialized to 0.
+#' - `objective` is initialized to `survival:aft`.
```
Ok I don't understand this. Why is there still the `objective` parameter when the learner sets it to aft?

sebffischer replied:
aft-nloglik is not documented

bblodfon replied:
Added the doc to `aft-nloglik`. Explained the other question in the below comment

sebffischer commented on R/learner_xgboost_surv_xgboost_cox.R:
```diff
@@ -0,0 +1,200 @@
+#' @description
+#' eXtreme Gradient Boosting regression using a **Cox Proportional Hazards**
+#' objective.
+#' Calls [xgboost::xgb.train()] from package \CRANpkg{xgboost} with `objective`
```
as above, I don't udnerstand why we add different learners for the objectives but then still expose the parameter

sebffischer replied:
also, can you remind me why we are refactoring this into two learners (I remember there is a good reason but I don't remember what it was ...)

bblodfon replied:
I couldn't make it work with just one learner, see a comment I wrote above, where I setup a `surv.xgboost.cox` learner with an `objective` parameter that pretty much changes the learner type which is confusing (so a Cox learner in name, but underneath an AFT fitting and so on)

sebffischer replied:
ok, but see my comment on mattermost

sebffischer commented on R/learner_xgboost_surv_xgboost_cox.R:
```diff
@@ -0,0 +1,200 @@
+#' - `nrounds` is initialized to 1.
+#' - `nthread` is initialized to 1 to avoid conflicts with parallelization via \CRANpkg{future}.
+#' - `verbose` is initialized to 0.
+#' - `objective` is initialized to `survival:cox`.
```
missing vals

bblodfon replied:
added `eval-metric`

sebffischer commented on R/learner_xgboost_surv_xgboost_cox.R:
```diff
@@ -142,8 +137,10 @@ LearnerSurvXgboostCox = R6Class("LearnerSurvXgboostCox",
     .train = function(task) {
       pv = self$param_set$get_values(tags = "train")
+      # manually add 'objective' and 'eval_metric'
+      pv = c(pv, objective = "survival:cox", eval_metric = "cox-nloglik")
```
if pv is empty I think this will construct a character vector and not a list

sebffischer replied:
Personally, I would not remove the eval metric (I might want to optimizer objective1 but also keep track of objective2)

sebffischer commented on R/learner_xgboost_surv_xgboost_cox.R:
```diff
@@ -169,6 +166,9 @@ LearnerSurvXgboostCox = R6Class("LearnerSurvXgboostCox",
     .predict = function(task) {
       pv = self$param_set$get_values(tags = "predict")
+      # manually add 'objective'
+      pv = c(pv, objective = "survival:cox")
```
same as above, (`c()` has some surprising properties :D)

sebffischer commented on R/learner_xgboost_surv_xgboost_aft.R:
```diff
@@ -171,6 +166,8 @@ LearnerSurvXgboostAFT = R6Class("LearnerSurvXgboostAFT",
     .predict = function(task) {
       pv = self$param_set$get_values(tags = "predict")
+      # manually add 'objective'
+      pv = c(pv, objective = "survival:aft")
```
see below

sebffischer commented on R/learner_xgboost_surv_xgboost_aft.R:
```diff
@@ -147,6 +140,8 @@ LearnerSurvXgboostAFT = R6Class("LearnerSurvXgboostAFT",
     .train = function(task) {
       pv = self$param_set$get_values(tags = "train")
+      # manually add 'objective' and 'eval_metric'
+      pv = c(pv, objective = "survival:aft", eval_metric = "aft-nloglik")
```
see below

sebffischer reviewed (commented):
minor details
